File: GolangSolutions/BinarySearch/Q69_x的平方根.go
package BinarySearch

import "math"

func mySqrt(x int) int {
	if x == 0 {
		return 0
	}
	ans := int(math.Exp(0.5 * math.Log(float64(x))))

	if (ans+1)*(ans+1) <= x {
		return ans + 1
	}
	return ans
}

func mySqrt2(x int) int {
	low, high := 0, x
	ans := -1
	for low <= high {
		mid := low + (high-low)/2
		if mid*mid <= x {
			ans = mid
			low = mid + 1
		} else {
			high = mid - 1
		}
	}

	return ans
}

File: LeetCode/src/BinarySearch/Q69_x的平方根.java
package BinarySearch;

/**
 * @author Hyper
 * @date 2022/10/10
 * <p>
 * 思路
 * 1. 数学做法，先转化为以e为底数再套用Math工具类解决
 * 2. 二分
 */
public class Q69_x的平方根 {
    public int mySqrt(int x) {
        if (x == 0) {
            return 0;
        }
        int ans = (int) Math.exp(0.5 * Math.log(x));
        return (long) (ans + 1) * (ans + 1) <= x ? ans + 1 : ans;
    }

    public int mySqrt2(int x) {
        int low = 0, high = x;
        int ans = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            // 需要注意溢出的情况
            if ((long) mid * mid <= x) {
                ans = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return ans;
    }
}
